WIP New hooks description (#586)
kubev2v/forklift-documentation: documentation/modules/adding-hook-using-cli.adoc

// Module included in the following assemblies:
//
// * documentation/doc-Migration_Toolkit_for_Virtualization/master.adoc

:_content-type: PROCEDURE
[id="adding-migration-hook-via-cli_{context}"]
= Adding a migration hook to a migration plan using the CLI

You can use a `Hook` CR to add a pre-migration hook or a post-migration hook to an existing migration plan using the {project-first} CLI.

For example, you can create a `Hook` CR to install the `cloud-init` service on a VM and write a file before migration.

[NOTE]
====

> You can run one pre-migration hook, one post-migration hook, or one of each per migration plan. Each hook needs its own `Hook` needs its own CR, but a `Plan` CR contains data for all the hooks it uses.

Reviewer: ```suggestion
You can run one pre-migration hook, one post-migration hook, or one of each per migration plan. Each hook needs its own `Hook` and its own CR, but a `Plan` CR contains data for all the hooks it uses.
```

Author: Fixed.

====

[NOTE]
====
You can retrieve additional information stored in a secret or in a `ConfigMap` by using a `k8s` module.
====

.Prerequisites

* Migration plan
* Migration hook image or the playbook containing the hook image
* File containing the Secret for the source provider
* {ocp} service account called by the hook and that has at least write access for the namespace you are working in

> [NOTE]
> ====
> For instructions for creating a service account, see link:https://docs.openshift.com/container-platform/{ocp-version}/authentication/understanding-and-creating-service-accounts.html
> ====

Reviewer: ```suggestion
.Additional information
For instructions for creating a service account, see link:https://docs.openshift.com/container-platform/{ocp-version}/authentication/understanding-and-creating-service-accounts.html
```
I do not think you should use an admonition for additional information

Author: Fixed.

Reviewer: No, it is not an admonition. Placement is irrelevant

It is `.Addtiional information`

Author: @anarnold97 Yes, I took off the admonition formatting, labeled it as "Additional resources", and placed it after the prerequisites. 

Should I try to reduce the number of admonitions?

> [NOTE]
> ====
> For instructions for creating a service account, see link:https://docs.openshift.com/container-platform/{ocp-version}/authentication/understanding-and-creating-service-accounts.html
> ====

Reviewer: ```suggestion
+
For instructions for creating a service account, see link:https://docs.openshift.com/container-platform/{ocp-version}/authentication/understanding-and-creating-service-accounts.html
```

* SSH access for VMs you want to migrate with the public key installed on the VMs
* VMs running on Microsoft Server only: Remote Execution enabled

.Procedure
. If needed, create a Secret with an SSH private key for the VM.
.. Choose an existing key or generate a key pair.
.. Install the public key on the VM.
.. Encode the private key in the Secret to base64.
+
[source,yaml]
----
apiVersion: v1
data:
  key: VGhpcyB3YXMgZ2Vu...
kind: Secret
metadata:
  name: ssh-credentials
  namespace: openshift-mtv
type: Opaque
----

> . Encode your playbook by concatenating a file and piping it for base64, for example:

Reviewer: ```suggestion
. Encode your playbook by concatenating a file and piping it for Base64 encoding, for example:
```

Author: Fixed.

// Module included in the following assemblies:
//
// * documentation/doc-Migration_Toolkit_for_Virtualization/master.adoc

:_content-type: PROCEDURE
[id="adding-migration-hook-via-cli_{context}"]
= Adding a migration hook to a migration plan using the CLI

You can add a migration hook to an existing migration plan using the {project-first} CLI.

> You can add a `PreHook` or a `PostHook` Hook CR when you migrate a virtual machine from the command line by using the {project-full} API. A `PreHook` runs before a migration, a `PostHook`, after.

Reviewer: ```suggestion
You can add a `PreHook` or a `PostHook` Hook CR when you migrate a virtual machine from the command line by using the {project-full} API. A `PreHook` runs before a migration, a `PostHook`, after a migration.
```
Sometimes best to be explicit about this type of thing

[NOTE]
====
You can retrieve additional information stored in a secret or in a `configMap` by using a `k8s` module.
====

> For example, you can create a hook CR to install `cloud-init` on a VM and write a file before migration.

Reviewer: ```suggestion
For example, you can create a hook CR to install the `cloud-init` service on a VM and write a file before migration.
```

.Prerequisites

* Migration plan
* Migration hook image or the playbook containing the hook image

[NOTE]
====
You can run one pre-migration hook, one post-migration hook, or one of each, per migration plan.
====

* File containing the `Secret` for the source provider

> * {ocp} service account that has at least write-access permission for the namespace you are working in

Reviewer: ```suggestion
* {ocp} service account that has at least write access for the namespace you are working in
```

* SSH access for VMs you want to migrate with the public key installed on the VMs
* VMs running on Microsoft Server only: Remote Execution enabled

.Procedure

> . If needed, create a Secret with an SSH private key for the VM. You can either use an existing key or generate a key pair, install the public key on the VM, and base64 encode the private key in the Secret.

Reviewer 2: Maybe: If needed, create a Secret with an SSH private key for the VM.

1. (sub-steps) Choose an existing key or generate a key pair.
2. Install the public key on the VM.
3. Encode the private key in the secret to base64.

+
[source,yaml]
----
apiVersion: v1
data:

>   key: VGhpcyB3YXMgZ2VuZXJhdGVkIHdpdGggc3NoLWtleWdlbiBwdXJlbHkgZm9yIHRoaXMgZXhhbXBsZS4KSXQgaXMgbm90IHVzZWQgYW55d2hlcmUuCi0tLS0tQkVHSU4gT1BFTlNTSCBQUklWQVRFIEtFWS0tLS0tCmIzQmxibk56YUMxclpYa3RkakVBQUFBQUJHNXZibVVBQUFBRWJtOXVaUUFBQUFBQUFBQUJBQUFCbHdBQUFBZHpjMmd0Y24KTmhBQUFBQXdFQUFRQUFBWUVBMzVTTFRReDBFVjdPTWJQR0FqcEsxK2JhQURTTVFuK1NBU2pyTGZLNWM5NGpHdzhDbnA4LwovRHErZHFBR1pxQkg2ZnAxYmVJM1BZZzVWVDk0RVdWQ2RrTjgwY3dEcEo0Z1R0NHFUQ1gzZUYvY2x5VXQyUC9zaTNjcnQ0CjBQdi9wVnZXU1U2TlhHaDJIZC93V0MwcGh5Z0RQOVc5SHRQSUF0OFpnZmV2ZnUwZHpraVl6OHNVaElWU2ZsRGpaNUFqcUcKUjV2TVVUaGlrczEvZVlCeTdiMkFFSEdzYU8xN3NFbWNiYUlHUHZuUFVwWmQrdjkyYU1JdWZoYjhLZkFSbzZ3Ty9ISW1VbQovdDdHWFBJUmxBMUhSV0p1U05odTQzZS9DY3ZYd3Z6RnZrdE9kYXlEQzBMTklHMkpVaURlNWd0UUQ1WHZXc1p3MHQvbEs1CklacjFrZXZRNUJsYWNISmViV1ZNYUQvdllpdFdhSFo4OEF1Y0czaGh2bjkrOGNSTGhNVExiVlFSMWh2UVpBL1JtQXN3eE0KT3VJSmRaUmtxTThLZlF4Z28zQThRNGJhQW1VbnpvM3Zwa0FWdC9uaGtIOTRaRE5rV2U2RlRhdThONStyYTJCZkdjZVA4VApvbjFEeTBLRlpaUlpCREVVRVc0eHdTYUVOYXQ3c2RDNnhpL1d5OURaQUFBRm1NRFBXeDdBejFzZUFBQUFCM056YUMxeWMyCkVBQUFHQkFOK1VpMDBNZEJGZXpqR3p4Z0k2U3RmbTJnQTBqRUova2dFbzZ5M3l1WFBlSXhzUEFwNmZQL3c2dm5hZ0JtYWcKUituNmRXM2lOejJJT1ZVL2VCRmxRblpEZk5ITUE2U2VJRTdlS2t3bDkzaGYzSmNsTGRqLzdJdDNLN2VORDcvNlZiMWtsTwpqVnhvZGgzZjhGZ3RLWWNvQXovVnZSN1R5QUxmR1lIM3IzN3RIYzVJbU0vTEZJU0ZVbjVRNDJlUUk2aGtlYnpGRTRZcExOCmYzbUFjdTI5Z0JCeHJHanRlN0JKbkcyaUJqNzV6MUtXWGZyL2RtakNMbjRXL0Nud0VhT3NEdnh5SmxKdjdleGx6eUVaUU4KUjBWaWJrallidU4zdnduTDE4TDh4YjVMVG5Xc2d3dEN6U0J0aVZJZzN1WUxVQStWNzFyR2NOTGY1U3VTR2E5WkhyME9RWgpXbkJ5WG0xbFRHZy83MklyVm1oMmZQQUxuQnQ0WWI1L2Z2SEVTNFRFeTIxVUVkWWIwR1FQMFpnTE1NVERyaUNYV1VaS2pQCkNuME1ZS053UEVPRzJnSmxKODZONzZaQUZiZjU0WkIvZUdRelpGbnVoVTJydkRlZnEydGdYeG5Iai9FNko5UTh0Q2hXV1UKV1FReEZCRnVNY0VtaERXcmU3SFF1c1l2MXN2UTJRQUFBQU1CQUFFQUFBR0JBSlZtZklNNjdDQmpXcU9KdnFua2EvakRrUwo4TDdpSE5mekg1TnRZWVdPWmRMTlk2L0lRa1pDeFcwTWtSKzlUK0M3QUZKZzBNV2Q5ck5PeUxJZDkxNjZoOVJsNG0xdFJjCnViZ1o2dWZCZ3hGVDlXS21mSEdCNm4zelh5b2pQOEFJTnR6ODVpaUVHVXFFRWtVRVdMd0RGSmdvcFllQ3l1VmZ2ZE92MUgKRm1WWmEwNVo0b3NQNkNENXVmc2djQ1RYQTR6VnZ5ZHVCYkxqdHN5RjdYZjNUdjZUQ1QxU0swZHErQk1OOXRvb0RZaXpwagpzbDh6NzlybXp3eUFyWFlVcnFUUkpsNmpwRkNrWHJLcy9LeG96MHhhbXlMY2RORk9hWE51LzlnTkpjRERsV2hPcFRqNHk4CkpkNXBuV1Jueis1RHJLRFdhY0loUW1CMUxVd2ZLWmQwbVFxaUpzMUMxcXZVUmlKOGExaThKUTI4bHFuWTFRRk9wbk13emcKWEpla2FndThpT1ExRFJlQkhaM0NkcVJUYnY3bVJZSGxramx0dXJmZGc4M3hvM0ErZ1JSR001eUVOcW5xSkplQjhJQVB5UwptMFp0dGdqbHNqNTJ2K1B1NmExMHoxZndKK1VML2N6dTRKeEpOYlp6WTFIMnpLODJBaVI1T3JYNmx2aUEvSWFSRVcwUUFBCkFNQndVeUJpcUc5bEZCUnltL2UvU1VORVMzdHpicUZNdTdIcy84WTV5SnAxKzR6OXUxNGtJR2ttV0Y5eE5HT3hrY3V0cWwKeHVUcndMbjFUaFNQTHQrTjUwTGhVdzR4ZjBhNUxqemdPbklPU0FRbm5HY1Nxa0dTRDlMR21obGE2WmpydFBHY29lQ3JHdAo5M1Vvcmx5YkxNRzFFRFAxWmpKS1RaZzl6OUMwdDlTTGd3ei9DbFhydW9UNXNQVUdKWnUrbHlIZXpSTDRtcHl6OEZMcnlOCkdNci9leVM5bWdISjNVVkZEYjNIZ3BaK1E1SUdBRU5rZVZEcHIwMGhCZXZndGd6YWtBQUFEQkFQVXQ1RitoMnBVby94V1YKenRkcVQvMzA4dFB5MXVMMU1lWFoydEJPQmRwSDJyd0JzdWt0aTIySGtWZUZXQjJFdUlFUXppMzY3MGc1UGdxR1p4Vng4dQpobEE0Rkg4ZXN1NTNQckZqVW9EeFJhb3d3WXBFcFh5Y2pnNUE1MStwR1VQcWljWjB0YjliaWlhc3BWWXZhWW5sdGlnVG5iClN0UExMY29nemNiL0dGcVYyaXlzc3lwTlMwKzBNRTUxcEtxWGNaS2swbi8vVHpZWWs4TW8vZzRsQ3pmUEZQUlZrVVM5blIKWU1pQzRlcEk0TERmbVdnM0xLQ2N1Zk85all3aWgwYlFBQUFNRUE2WEtldDhEMHNvc0puZVh5WFZGd0dyVyszNlhBVGRQTwpMWDdjaStjYzFoOGV1eHdYQWx3aTJJNFhxSmJBVjBsVEhuVGEycXN3Uy9RQlpJUUJWSkZlVjVyS1daZTc4R2F3d1pWTFZNCldETmNwdFFyRTFaM2pGNS9TdUVzdlVxSDE0Tkc5RUFXWG1iUkNzelE0Vlk3NzQrSi9sTFkvMnlDT1diNzlLYTJ5OGxvYUoKVXczWWVtSld3blp2R3hKNldsL3BmQ2xYN3lEVXlXUktLdGl0cWNjbmpCWVkyRE1tZURwdURDYy9ZdDZDc3dLRmRkMkJ1UwpGZGt5cDlZY3VMaDlLZEFBQUFIR3BoYzI5dVFFRlVMVGd3TWxVdWJXOXVkR3hsYjI0dWFXNTBjbUVCQWdNRUJRWT0KLS0tLS1FTkQgT1BFTlNTSCBQUklWQVRFIEtFWS0tLS0tCgo=

Reviewer: Not sure we need to include this `secret`.

Do you think we can abbreviate?

Reviewer: ```suggestion
  key: VGhpcyB3YXMgZ2Vu ... 
```

Author: Good idea.

kind: Secret
metadata:
  name: ssh-credentials
  namespace: konveyor-forklift
type: Opaque
----
. Encode your playbook by concatenating a file and piping it for base64, for example:
+
[source,shell]
----
$ cat playbook.yml | base64 -w0
----
+
[NOTE]
====

> You can also use a `here` document to encode a playbook:

Reviewer: could you explain a bit more, as i got a bit confused?

Author: @anarnold97 I'm actually wondering if this is still relevant -- it was in the upstream document that I refashioned into the current downstream documentation. I'll make a note to ask Martin and Matthew about it when I ask for their review.

Author: I did change the wording.

Author: @mnecas Is this note about a here document still relevant?

Author: Note removed, users don't really need it.

[NOTE]
====
You can run one pre-migration hook, one post-migration hook, or one of each per migration plan. Each hook needs its own `Hook` needs its own CR, but a `Plan` CR contains data for all the hooks it uses.
====

[NOTE]
====
You can retrieve additional information stored in a secret or in a `ConfigMap` by using a `k8s` module.
====

.Prerequisites

* Migration plan
* Migration hook image or the playbook containing the hook image

[NOTE]
====
You can run one pre-migration hook, one post-migration hook, or one of each per migration plan.
====

* File containing the Secret for the source provider
* {ocp} service account that has at least write access for the namespace you are working in
* SSH access for VMs you want to migrate with the public key installed on the VMs
* VMs running on Microsoft Server only: Remote Execution enabled

.Procedure
. If needed, create a Secret with an SSH private key for the VM.
.. Choose an existing key or generate a key pair.
.. Install the public key on the VM.
.. Encode the private key in the Secret to base64.
+
[source,yaml]
----
apiVersion: v1
data:
  key: VGhpcyB3YXMgZ2Vu...
kind: Secret
metadata:
  name: ssh-credentials

>   namespace: konveyor-forklift

Reviewer 3: lets use the `openshift-mtv` namespace as default in the product docs.

type: Opaque
----
. Encode your playbook by concatenating a file and piping it for base64, for example:
+
[source,shell]
----
$ cat playbook.yml | base64 -w0
----
+
[NOTE]
====
You can also use a `here` document to add instructions to encode a playbook inside a larger command:

[source,shell]
----
$ cat << EOF | base64 -w0
- hosts: localhost
  tasks:
  - debug:
      msg: test
EOF
----
====

> 

Author: @mnecas Do you have a better way to explain what a 'here" document is? Is this note actually needed/useful for our users?

Reviewer 3: yeah the "here" document is the `EOF` at the start and the end 
https://tldp.org/LDP/abs/html/here-docs.html

. Create a Hook CR:
+
[source,yaml,subs="attributes+"]
----
$  cat << EOF | {oc} apply -f -
apiVersion: forklift.konveyor.io/v1beta1
kind: Hook
metadata:
  name: <hook>
  namespace: <namespace>
spec:
  image: quay.io/konveyor/hook-runner
  playbook: |
    LS0tCi0gbm...
EOF
----
+
where:
+
`playbook` refers to an optional Base64-encoded Ansible Playbook. If you specify a playbook, the `image` must be `hook-runner`.

> +

Author: @mnecas @matthewsecaur comments:
f we're going to create a Hook CR here, then we may as well have the user put the serviceAccount in now, instead of creating the hook and then adding the SA later. In fact, I don't see any mention of the serviceAccount in the CLI-based instructions. This is required for Console-based and CLI-based hooks.
The example command for decoding a playbook from a hook uses the "konveyor-forklift" namespace, but MTV is installed in "openshift-mtv"
Same as earlier, the text says "If you specify a playbook, the image must be hook-runner", but that isn't necessarily true. You can use a custom image *and* a playbook.

Reviewer 3: yeah, we should create the hook with the `serviceAccount` the problem is that the latest UI does not support it so as workaround we needed to update it manually inside the CR yaml.

Reviewer 3: "If you specify a playbook, the image must be hook-runner" it does not need to be a hook-runner but it needs to have a `ansible-runner`  as we run that user specific playbook in
https://github.com/kubev2v/forklift/blob/0429598fe82ff331ab65e72ff391d1cbb3e1c3b4/pkg/controller/plan/hook.go#L225-L235

If the user does not specify the playbook we will run the image itself as user configured it.

Author: @mnecas Sounds like I need to add serviceAccount to the CLI instructions -- let's talk about how to do that.

* VMs running on Microsoft Server only: Remote Execution enabled

.Procedure
. If needed, create a Secret with an SSH private key for the VM.
.. Choose an existing key or generate a key pair.
.. Install the public key on the VM.
.. Encode the private key in the Secret to base64.
+
[source,yaml]
----
apiVersion: v1
data:
  key: VGhpcyB3YXMgZ2Vu...
kind: Secret
metadata:
  name: ssh-credentials
  namespace: openshift-mtv
type: Opaque
----
. Encode your playbook by concatenating a file and piping it for base64, for example:
+
[source,shell]
----
$ cat playbook.yml | base64 -w0
----
+
[NOTE]
====
You can also use a `here` document - a file preceded and followed by `EOF` - to add instructions to encode a playbook inside a larger command:

.Example

[source,shell]
----
$ cat << EOF | base64 -w0
- hosts: localhost
  tasks:
  - debug:
      msg: test
EOF
----
====

. Create a Hook CR:
+
[source,yaml,subs="attributes+"]
----
$  cat << EOF | {oc} apply -f -
apiVersion: forklift.konveyor.io/v1beta1
kind: Hook
metadata:
  name: <hook>
  namespace: <namespace>
spec:
  image: quay.io/konveyor/hook-runner
  playbook: |

>     LS0tCi0gbm...

Reviewer 3: Example:
```yaml
apiVersion: forklift.konveyor.io/v1beta1
kind: Hook
metadata:
  name: <hook>
  namespace: <namespace>
spec:
  image: quay.io/konveyor/hook-runner
  serviceAccount: hook-account
  playbook: |
    LS0tCi0gbm...
```
We need to document why the users would need to have the `serviceAccount`.
The main reason for this is that if the user wants to manipulate any resources from the cluster, they need to specify the account with which it will log in to the cluster and manipulate the resource.

Reviewer 4: It's true that a serviceAccount isn't strictly required, but even to access something like a public key that's stored in an OCP secret, it needs to be provided. Either the user needs to supply a bunch of stuff in a custom hook image or they need to extract the data to access VMs or other resources in their environment, but the docs aren't very clear about what is needed in each of those cases. The simpler approach that is covered here (i.e. to use the default hook runner image) would require a serviceAccount in order to do anything useful at all.
